atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/monitoring/TimestampTrackerTest.java

```diff
@@ -0,0 +1,161 @@
+                    .contains(buildFullyQualifiedMetricName(FRESH_TIMESTAMP_NAME))
+                    .contains(buildFullyQualifiedMetricName(UNREADABLE_TIMESTAMP_NAME));
+        }
+    }
+
+    @Test
+    public void defaultTrackersDelegateToRelevantComponents() {
+        try (TimestampTracker ignored = TimestampTracker.createWithDefaultTrackers(timelockService, cleaner)) {
+            when(timelockService.getImmutableTimestamp()).thenReturn(ONE);
+            when(timelockService.getFreshTimestamp()).thenReturn(TEN);
+            when(cleaner.getUnreadableTimestamp()).thenReturn(FORTY_TWO);
+
+            assertThat(getGauge(IMMUTABLE_TIMESTAMP_NAME).getValue()).isEqualTo(ONE);
+            assertThat(getGauge(FRESH_TIMESTAMP_NAME).getValue()).isEqualTo(TEN);
+            assertThat(getGauge(UNREADABLE_TIMESTAMP_NAME).getValue()).isEqualTo(FORTY_TWO);
+
+            verify(timelockService).getImmutableTimestamp();
+            verify(timelockService).getFreshTimestamp();
+            verify(cleaner).getUnreadableTimestamp();
+            verifyNoMoreInteractions(timelockService, cleaner);
+        }
+    }
+
+    @Test
+    public void metricsAreDeregisteredUponClose() {
+        try (TimestampTracker tracker = new TimestampTracker(Clock.defaultClock())) {
+            tracker.registerTimestampForTracking(FAKE_METRIC, () -> 1L);
+            assertThat(AtlasDbMetrics.getMetricRegistry().getNames())
+                    .contains(buildFullyQualifiedMetricName(FAKE_METRIC));
+        }
+
+        assertThat(AtlasDbMetrics.getMetricRegistry().getNames())
+                .doesNotContain(buildFullyQualifiedMetricName(FAKE_METRIC));
+    }
+
+    @Test
+    public void canCloseMultipleTimes() {
+        TimestampTracker tracker = new TimestampTracker(Clock.defaultClock());
+        tracker.registerTimestampForTracking(FAKE_METRIC, () -> 1L);
+
+        tracker.close();
+        tracker.close();
+        assertThat(AtlasDbMetrics.getMetricRegistry().getNames())
+                .doesNotContain(buildFullyQualifiedMetricName(FAKE_METRIC));
+    }
+
+    @Test
+    public void doesNotCallSupplierOnRequestsWithinRetriggerInterval() {
+        try (TimestampTracker tracker = new TimestampTracker(mockClock)) {
+            when(mockClock.getTick()).thenReturn(0L, 1L, 2L);
+            AtomicLong timestampValue = new AtomicLong(0L);
+            tracker.registerTimestampForTracking(FAKE_METRIC, timestampValue::incrementAndGet);
+
+            assertThat(getGauge(FAKE_METRIC).getValue()).isEqualTo(1L);
+            assertThat(getGauge(FAKE_METRIC).getValue()).isEqualTo(1L);
+            assertThat(getGauge(FAKE_METRIC).getValue()).isEqualTo(1L);
+        }
+    }
+
+    @Test
+    public void callsSupplierAgainAfterTimeElapses() {
+        try (TimestampTracker tracker = new TimestampTracker(mockClock)) {
+            when(mockClock.getTick()).thenReturn(0L, CACHE_RETRIGGER_NANOS / 2, CACHE_RETRIGGER_NANOS);
```

Comment:
For a stronger test, use `CACHE_RETRIGGER_NANOS - 2`.

For a clearer test, set `CACHE_INTERVAL_NANOS = TimestampTracker.CACHE_INTERVAL.toNanos()`, and use `CACHE_INTERVAL_NANOS - 1` and `CACHE_INTERVAL_NANOS + 1`.

Reply:
The second alternative you propose does both - and yep, it is both stronger and more readable that way.